File: src/main/scala/org/lolhens/minechanics/core/asm/ClassTransformer.scala
package org.lolhens.minechanics.core.asm

import net.minecraft.launchwrapper.IClassTransformer

class ClassTransformer extends IClassTransformer {
  override def transform(name: String, deobfName: String, basicClass: Array[Byte]): Array[Byte] = {
    val transformation = Transformations.transformations.get(deobfName)
    if (transformation.nonEmpty) {
      val transformedClass = transformation.get.transform(basicClass)
      if (transformedClass != null) return transformedClass
    }
    basicClass
  }
}

File: src/main/java/org/lolhens/minechanics/core/asm/ClassTransformer.java
package org.lolhens.minechanics.core.asm;

import net.minecraft.launchwrapper.IClassTransformer;

import org.lolhens.minechanics.core.event.Events;

import com.dafttech.eventmanager.Event;
import com.dafttech.reflect.ReflectionUtil.SingletonInstance;

public class ClassTransformer implements IClassTransformer {
    @SingletonInstance
    public ClassTransformer instance;

    public ClassTransformer() {
        instance = this;
    }

    @Override
    public byte[] transform(String name, String deobfName, byte[] basicClass) {
        Event transformEvent = Events.eventManager.callSync(Events.transform, deobfName, new DeobfTranslator(name, deobfName),
                basicClass);
        if (!transformEvent.isCancelled())
            if (transformEvent.out.size() > 0) return (byte[]) transformEvent.out.get(transformEvent.out.size() - 1);
        return basicClass;
    }

    public static void registerASMListeners() {
        Events.eventManager.tryRegisterEventListener("org.lolhens.minechanics.liquids.LiquidTransformer");
    }
}